From 0d97cbdc186c6bd82f3caf71f5a69d830c1f383b Mon Sep 17 00:00:00 2001
Message-Id: <0d97cbdc186c6bd82f3caf71f5a69d830c1f383b.1425610844.git.chang-joon.lee@intel.com>
In-Reply-To: <6fa2ab23adf6d7de781e906d04a0224410262d9b.1425610844.git.chang-joon.lee@intel.com>
References: <6fa2ab23adf6d7de781e906d04a0224410262d9b.1425610844.git.chang-joon.lee@intel.com>
From: Kausal Malladi <Kausal.Malladi@intel.com>
Date: Mon, 16 Feb 2015 18:04:11 +0530
Subject: [PATCH 19/27] SQUASHME! [VPG]: video/adf: Fix for Color correction
 in Sprite registers

This patch adds a bug fix for Sprite register offset calculation.
Sprite registers are continuously placed at an offset of 0x100.
But the enum plane adds primary planes also in this sequence.

To compensate for that, this patch alters the plane id,
so that a direct sprite offset can be used in color calculations.

SQUASHME! - This patch should be squashed into the following existing patch:

	Author: Kausal Malladi <Kausal.Malladi@intel.com>
	Date:   Sun Feb 1 13:38:02 2015 +0530
	Change-Id: I13536ecf7954c7d74b1f038e08017cc332598f1a
	MUST_REBASE [VPG]: video/adf: Color Manager Framework Actions

Issue: GMINL-4163
Change-Id: I160e1287235fb20d4780871127c3609a8c4fbb20
Signed-off-by: Kausal Malladi <Kausal.Malladi@intel.com>
---
 drivers/video/adf/intel/intel_adf_overlay_engine.c |   27 ++++++++++++++++++--
 1 file changed, 25 insertions(+), 2 deletions(-)

diff --git a/drivers/video/adf/intel/intel_adf_overlay_engine.c b/drivers/video/adf/intel/intel_adf_overlay_engine.c
index af060e4..9693e58 100644
--- a/drivers/video/adf/intel/intel_adf_overlay_engine.c
+++ b/drivers/video/adf/intel/intel_adf_overlay_engine.c
@@ -21,7 +21,9 @@ long intel_overlay_engine_obj_ioctl(struct adf_obj *obj,
 	struct intel_adf_overlay_engine *overlay_engine;
 	struct intel_plane *plane;
 	struct adf_overlay_engine *adf_oe;
+	struct intel_pipe *pipe;
 	u8 plane_id;
+	u8 pipe_id;
 
 	long err = 0;
 
@@ -32,11 +34,23 @@ long intel_overlay_engine_obj_ioctl(struct adf_obj *obj,
 	plane = overlay_engine->plane;
 	plane_id = plane->base.idx;
 
+	pipe = plane->pipe;
+	pipe_id = pipe->base.idx;
+
 	switch (cmd) {
 	case INTEL_ADF_COLOR_MANAGER_SET:
 		pr_info("ADF: Calling apply to set Color Property on the Overlay Engine\n");
+
+		/*
+		 * Sprite registers are continuously placed at
+		 * an offset of 0x100. But the Plane Enum adds
+		 * primary planes also into account.
+		 * To compensate for that, alter the plane id,
+		 * so that a direct sprite offset can be used
+		 * in applying color correction.
+		 */
 		if (!intel_color_manager_apply(plane->color_ctx,
-				(struct color_cmd *) arg, plane_id)) {
+				(struct color_cmd *) arg, plane_id - pipe_id)) {
 			pr_err("%s Error: Set color correction failed\n",
 								__func__);
 			return -EFAULT;
@@ -49,8 +63,17 @@ long intel_overlay_engine_obj_ioctl(struct adf_obj *obj,
 
 	case INTEL_ADF_COLOR_MANAGER_GET:
 		pr_info("ADF: Calling get Color Property on the Overlay Engine\n");
+
+		/*
+		 * Sprite registers are continuously placed at
+		 * an offset of 0x100. But the Plane Enum adds
+		 * primary planes also into account.
+		 * To compensate for that, alter the plane id,
+		 * so that a direct sprite offset can be used
+		 * in applying color correction.
+		 */
 		if (!intel_color_manager_get(plane->color_ctx,
-				(struct color_cmd *) arg, plane_id)) {
+				(struct color_cmd *) arg, plane_id - pipe_id)) {
 			pr_err("%s Error: Get color correction failed\n",
 								__func__);
 			return -EFAULT;
-- 
1.7.9.5

